File: android/ValleTPV/ValleTPVlib/src/main/java/com/valleapp/valletpvlib/tareas/TareaManejarInstrucciones.java
package com.valleapp.valletpv.tareas;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.support.annotation.NonNull;
import android.util.Log;


import com.valleapp.valletpv.tools.Instrucciones;
import com.valleapp.valletpvlib.comunicacion.HTTPRequest;

import java.util.Queue;
import java.util.TimerTask;

public class TareaManejarInstrucciones extends TimerTask {

    private final Queue<Instrucciones> cola;
    private final long timeout;
    boolean procesado = true;
    int count = 0;

    Handler handleHttp = new Handler(Looper.myLooper()){
        @Override
        public void handleMessage(@NonNull Message msg) {
            try {
                String op = msg.getData().getString("op");
                if (op != null && op.equals("ERROR")) {
                    cola.poll();
                }else {
                    String res = msg.getData().getString("RESPONSE");
                    if (res != null && !res.equals("")) {
                        synchronized (cola) {
                            Instrucciones inst = cola.poll();
                            if (inst != null) {
                                Handler handlerInst = inst.getHandler();
                                if (handlerInst != null) {
                                    Message msgInst = handlerInst.obtainMessage();
                                    Bundle bundle = msg.getData();
                                    if (bundle == null) bundle = new Bundle();
                                    bundle.putString("RESPONSE", res);
                                    bundle.putString("op", inst.getOp());
                                    msgInst.setData(bundle);
                                    handlerInst.sendMessage(msgInst);
                                }
                            }
                        }
                    }
                }
            }catch (Exception e){
                Log.w("Instrucciones", "Error al ejecutar instruccion en el servidor");
            }
            procesado = true;count=0;
            super.handleMessage(msg);

        }
    };

    public TareaManejarInstrucciones(Queue<Instrucciones> colaInstrucciones, long timeout) {
        this.cola= colaInstrucciones;
        this.timeout = timeout;
    }

    @Override
    public void run() {
        try {

            if (procesado) {
                synchronized (cola) {
                    Instrucciones inst = cola.peek();
                    if (inst != null) {
                        new HTTPRequest(inst.getUrl(), inst.getParams(), "", handleHttp);
                        procesado = false;
                    }
                }
            }else{
                count ++;
            }

            if (count > 20) {
                count = 0;
                procesado = true;
            }

            Thread.sleep(timeout);

        }catch (Exception e){
            e.printStackTrace();
        }
    }
}

File: android/ValleTPV/ValleTPVlib/src/main/java/com/valleapp/valletpvlib/tareas/TareaManejarInstrucciones.kt
package com.valleapp.valletpvlib.tareas

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.Log
import com.valleapp.valletpvlib.comunicacion.HTTPRequest
import com.valleapp.valletpvlib.tools.Instruccion
import java.util.Queue
import kotlin.random.Random

class TareaManejarInstrucciones(
    private val cola: Queue<Instruccion>,
    private var uid: String
) {

    private val handler = Handler(Looper.getMainLooper())
    private var procesando = false
    private var esperandoReintento = false
    private var tiempoReintento = 10000L // 10 segundos iniciales
    private val TIMEOUT_RESPONSE = 40000L // 40 segundos para esperar respuesta de HTTPRequest

    fun iniciar() {
        if (!procesando && !esperandoReintento) {
            procesarSiguiente()
        }
    }

    private val handleHttp = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            procesando = false // Liberamos el procesamiento al recibir mensaje

            try {
                val op = msg.data?.getString("op") ?: "unknown_op" // Evitamos NPE
                val res = msg.data?.getString("RESPONSE")
                val codigoEstado = msg.data?.getInt("codigoEstado") ?: -1
                val connectionStatus = msg.data?.getString("connectionStatus")

                when (connectionStatus) {
                    "server_unreachable" -> {
                        // Caso 1: Fallo de conexión, reintentamos indefinidamente
                        Log.w("TareaInstrucciones", "No se pudo conectar ($op), reintentando...")
                        esperarReintento()
                    }
                    "server_response" -> {
                        // El servidor respondió (éxito o error), avanzamos
                        synchronized(cola) {
                            val inst = cola.poll() // Eliminamos la instrucción
                            inst?.handler?.let { handlerInst ->
                                val msgInst = handlerInst.obtainMessage()
                                val bundle = msg.data ?: Bundle()
                                bundle.putString("RESPONSE", res)
                                bundle.putString("op", inst.op)
                                msgInst.data = bundle
                                handlerInst.sendMessage(msgInst)
                            }
                        }
                        if (codigoEstado >= 400) {
                            Log.e("TareaInstrucciones", "Error del servidor ($op): Código $codigoEstado, Respuesta: $res")
                        }
                        tiempoReintento = 10000L
                        procesarSiguiente()
                    }
                    else -> {
                        // Caso 4: Estado desconocido, descartamos y avanzamos
                        Log.e("TareaInstrucciones", "Estado de conexión inválido ($op): $connectionStatus")
                        synchronized(cola) { cola.poll() }
                        procesarSiguiente()
                    }
                }
            } catch (e: Exception) {
                // Caso 3: Excepción al procesar, descartamos y avanzamos
                val op = msg.data?.getString("op") ?: "unknown_op"
                Log.e("TareaInstrucciones", "Error al procesar mensaje ($op): ${e.message}")
                synchronized(cola) { cola.poll() }
                procesarSiguiente()
            }
        }
    }

    private fun procesarSiguiente() {
        synchronized(cola) {
            if (cola.isNotEmpty()) {
                procesando = true // Marcamos como procesando
                val inst = cola.peek() // Miramos la instrucción sin eliminarla
                inst?.let {
                    it.params.put("uid", uid)
                    println("Enviando instrucción: ${it.op} con params: ${it.params}, url: ${it.url}")

                    HTTPRequest(it.url, it.params, it.op, handleHttp) // Enviamos la solicitud

                    // Caso 2: Timeout si HTTPRequest no responde
                    handler.postDelayed({
                        if (procesando) { // Si aún no hemos recibido respuesta
                            Log.w("TareaInstrucciones", "Timeout esperando respuesta ($it.op), reintentando...")
                            procesando = false
                            esperarReintento() // Reintentamos como fallo de conexión
                        }
                    }, TIMEOUT_RESPONSE)
                }
            }
        }
    }

    private fun esperarReintento() {
        esperandoReintento = true
        val jitter = Random.nextLong(-1000, 1000) // Jitter para evitar colisiones
        handler.postDelayed({
            esperandoReintento = false
            tiempoReintento = (tiempoReintento * 2).coerceAtMost(60000) + jitter // Backoff exponencial, máximo 1 minuto
            procesarSiguiente()
        }, tiempoReintento)
        Log.w("TareaInstrucciones", "Reintentando, espera ${tiempoReintento / 1000} segundos")
    }

    fun setUid(uid: String?) {
        this.uid = uid ?: ""
    }
}